### PR TITLE
[BUG] Loosen random_state input restriction

### LINKED ISSUE #32503: core.common.random_state is too strict for state input
The numpy RandomState seed can be an int, array-like, or BitGenerator

```python
seed : {None, int, array_like, BitGenerator}, optional
    Random seed used to initialize the pseudo-random number generator or
    an instantized BitGenerator.  If an integer or array, used as a seed for
    the MT19937 BitGenerator. Values can be any integer between 0 and
    2**32 - 1 inclusive, an array (or other sequence) of such integers,
    or ``None`` (the default).  If `seed` is ``None``, then the `MT19937`
    BitGenerator is initialized by reading data from ``/dev/urandom``
    (or the Windows analogue) if available or seed from the clock
    otherwise.
```

Right now, we only pass through integers. Arrays (and BitGenerators if possible) should be passed through to RandomState at https://github.com/pandas-dev/pandas/blob/54c5e9e6e0819f848a10045d1cbc8ee936ffd075/pandas/core/common.py#L409-L410. 

## Current Behavior

```pytb
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: state_data = np.random.randint(0, 2**31, size=624, dtype='uint32')

In [4]: pd.core.common.random_state(state_data)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-4f97e41a5899> in <module>
----> 1 pd.core.common.random_state(state_data)

~/sandbox/pandas/pandas/core/common.py in random_state(state)
    415     else:
    416         raise ValueError(
--> 417             "random_state must be an integer, a numpy RandomState, or None"
    418         )
    419

ValueError: random_state must be an integer, a numpy RandomState, or None
```

## Expected Output

```python
Out[4]: RandomState(MT19937) at 0x115B16050
```

### PR DESCRIPTION
**Pull request by mikekutzma**
Alllow for array-like as well as BitGenerator inputs

Addresses: GH32503

- [x] closes #32503 
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/common.py [hunk 1]
```diff
@@ -406,7 +406,11 @@ def random_state(state=None):
     -------
     np.random.RandomState
     """
-    if is_integer(state):
+    if (
+        is_integer(state)
+        or is_array_like(state)
+        or isinstance(state, np.random.BitGenerator)
```
Comment (TomAugspurger) [inline pandas/core/common.py]: BitGenerator is somewhat new (NumPy>=1.17 IIRC). You'll need to put this in a conditional check. `pandas.compat` has some helpers.

**Review by TomAugspurger — COMMENTED** [review]
Can you add a release note to v1.1.0.rst?

Need to update the docstring at https://github.com/pandas-dev/pandas/pull/32510/files#diff-21cb6655a9828d526610249e3d6cb16eR398, and make sure to include a `.. versionmodified:: 1.1.0` noting that support for arrays and BitGenerators was added in 1.1.

File: pandas/core/common.py [hunk 2]
```diff
@@ -405,8 +407,16 @@ def random_state(state=None):
...
     np.random.RandomState
+
+    ..versionchanged:: 1.1.0
+        array-like and BitGenerator (for NumPy>=1.17) object now passed to
+        np.random.RandomState() as seed
     """
-    if is_integer(state):
+    if (
+        is_integer(state)
+        or is_array_like(state)
```
Comment (xcz011) [inline pandas/core/common.py]: `np.random.RandomState([1, 2.5, 3])` will throw `TypeError: Cannot cast array from dtype('float64') to dtype('int64') according to the rule 'safe'`. So I think just `is_array_like` won't be enough.

Comment (TomAugspurger) [reply pandas/core/common.py]: Tht's probably OK though. The traceback will point to the relevant section in NumPy, and if we want to improve that error message we can improve it in NumPy.

**Review by jreback — CHANGES_REQUESTED** [review]
can you also update the doc-string for .sample in pandas.core.generic

and also update a test to use a BItGenerator (grep for where these tests are)

I would actually change the whatsnew to mention .sample which is much more user facing.

File: pandas/tests/generic/test_generic.py [hunk 3]
```diff
@@ -654,6 +657,24 @@ def test_sample(sel):
         with pytest.raises(ValueError):
             df.sample(1, weights=s4)
 
+        # Check that random_state arguments are processed correctly
```
Comment (jreback) [inline pandas/tests/generic/test_generic.py]: can you make this a separate test & parameterize

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/generic/test_generic.py (hunk 3 above)
Comment (mikekutzma) [reply pandas/tests/generic/test_generic.py]: Done :+1:

File: pandas/core/generic.py [hunk 4]
```diff
@@ -4794,9 +4794,10 @@ def sample(
             If weights do not sum to 1, they will be normalized to sum to 1.
             Missing values in the weights column will be treated as zero.
             Infinite values not allowed.
-        random_state : int or numpy.random.RandomState, optional
-            Seed for the random number generator (if int), or numpy RandomState
-            object.
+        random_state : int, array-like, BitGenerator, np.random.RandomState, optional
+            If int, array-like, or BitGenerator (NumPy>=1.17), seed for
+            random number generator
+            If np.random.RandomState, use as numpy RandomState object.
```
Comment (jreback) [inline pandas/core/generic.py]: can you add versionchanged here

File: pandas/core/common.py [hunk 5]
```diff
@@ -405,16 +407,27 @@ def random_state(state=None):
     Returns
     -------
     np.random.RandomState
+
+    ..versionchanged:: 1.1.0
```
Comment (jreback) [inline pandas/core/common.py]: this needs to be in the Paramters section, just below the 'If receives anything'

File: pandas/core/generic.py (hunk 4 above)
Comment (jreback) [reply pandas/core/generic.py]: same as you have above (in the good location)

File: pandas/tests/generic/test_generic.py [hunk 6]
```diff
@@ -641,6 +644,30 @@ def test_sample(sel):
...
+                11,
+                marks=pytest.mark.skipif(_np_version_under1p17, reason="NumPy<1.17"),
+            ),
+        ],
+    )
+    def test_sample_random_state(self, func_str, arg):
+        # GH32503
+        df = pd.DataFrame({"col1": range(10, 20), "col2": range(20, 30)})
+        tm.assert_frame_equal(
+            df.sample(n=3, random_state=com.random_state(eval(func_str)(arg))),
```
Comment (jreback) [inline pandas/tests/generic/test_generic.py]: can you do

result =
expected = 
tm.assert_frame_equal(result, expected)

File: pandas/tests/test_common.py [hunk 7]
```diff
@@ -59,8 +61,29 @@ def test_random_state():
...
 
+    # check array-like
+    state_arr_like = npr.randint(0, 2 ** 31, size=624, dtype="uint32")
+    assert (
+        com.random_state(state_arr_like).uniform()
+        == npr.RandomState(state_arr_like).uniform()
+    )
+
+    # Check BitGenerators
+    if not _np_version_under1p17:
```
Comment (jreback) [inline pandas/tests/test_common.py]: can you add the issue number here as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, some doc-string changes, ping on green.

**Review by jreback — APPROVED** [review]

**Pull request merged**